Mathlib/Geometry/Convex/Cone/ConicalHull.lean

```diff
@@ -0,0 +1,200 @@
+
+section PointedConeSection
+variable [IsOrderedRing R]
+
+/-- 'IsPointedCone' - Predicate stating that a subset C of a monoid E with an R-module
+structure is a pointed cone. -/
+def IsPointedCone (C : Set E) : Prop :=
+  ∃ (T : PointedCone R E), (T.carrier = C)
+
+/-- Membership in PointedCone.span is equivalent to being a conical combination. -/
+theorem mem_span_iff_conicalCombination (s : Set E) (x : E) :
+    x ∈ PointedCone.span R s ↔ IsConicalCombination R s x := by
+  constructor
+  · intro hx
+    rw [Finsupp.mem_span_iff_linearCombination] at hx
+    rcases hx with ⟨l, rfl⟩
+    refine ⟨l.mapRange (↑) (by simp), (fun v => (l v).property), ?_⟩
+    classical
+    simp [Finsupp.linearCombination_apply, Finsupp.sum_mapRange_index]
+  · rintro ⟨f, hf_nonneg, rfl⟩
+    classical
+    rw [Finsupp.linearCombination_apply]
+    apply AddSubmonoid.sum_mem
+    intro v hv
+    exact (PointedCone.span R s).smul_mem' ⟨f v, hf_nonneg v⟩ (PointedCone.subset_span v.property)
+
+/-- Corollary: The PointedCone.span equals the set of all conical combinations. -/
+theorem span_eq_setOf_conicalCombination (s : Set E) :
+    (PointedCone.span R s : Set E) = {x | IsConicalCombination R s x} := by
+  ext x
+  exact mem_span_iff_conicalCombination R s x
+
+/-- A pointed cone contains all conical combinations of elements from any given subset. -/
+theorem conicalCombination_mem (s : Set E) (f : s →₀ R) (h_nonneg : ∀ v, 0 ≤ f v)
+    (C : PointedCone R E) (hs : s ⊆ C.carrier) :
+    (Finsupp.linearCombination R (Subtype.val : s → E)) f ∈ C := by
+  rw [Finsupp.linearCombination_apply]
+  apply AddSubmonoid.sum_mem
+  intro v hv
+  exact C.smul_mem' ⟨f v, h_nonneg v⟩ (hs v.property)
+
+/-- PointedCone.span R s is contained in any pointed cone C that contains s. -/
+theorem span_minimal (s : Set E) (C : PointedCone R E) (hs : s ⊆ C.carrier) :
+    (PointedCone.span R s : Set E) ⊆ C.carrier := by
+  intro x hx
+  have := (mem_span_iff_conicalCombination R s x).mp hx
+  obtain ⟨f, hf_nonneg, hf_eq⟩ := this
+  rw [hf_eq]
+  exact conicalCombination_mem R s f hf_nonneg C hs
+
+/-- PointedCone.span R s is equal to the infimum of all pointed cones containing s.
+This characterizes span as the smallest pointed cone containing s. -/
+theorem span_eq_sInf (s : Set E) :
+    PointedCone.span R s = sInf {C : PointedCone R E | s ⊆ C.carrier} := by
+  apply le_antisymm
+  · -- span R s ≤ sInf {C | s ⊆ C}
+    apply le_sInf
+    intro C hC
+    exact span_minimal R s C hC
+  · -- sInf {C | s ⊆ C} ≤ span R s
+    apply sInf_le
+    exact PointedCone.subset_span
+
+/-- C is a pointed cone iff C is nonempty and contains all two element conical combinations.
+This characterization of pointed cones is convenient in some proofs -/
+theorem pointedCone_iff_forall_nonneg {E : Type*} [AddCommMonoid E] [Module R E]
+    (C : Set E) : IsPointedCone R C ↔
+      C.Nonempty ∧ ∀ (x : E), x ∈ C → ∀ (y : E), y ∈ C → ∀ (a b : R),
+      0 ≤ a → 0 ≤ b → a • x + b • y ∈ C where
```

Comment:
We shouldn't have this `IsPointedCone` predicate. Instead, we should have a constructor for a pointed cone from `C.Nonempty ∧ ∀ (x : E), x ∈ C → ∀ (y : E), y ∈ C → ∀ (a b : R), 0 ≤ a → 0 ≤ b → a • x + b • y ∈ C`.
Something like
```
@[simps]
def PointedCone.of_coneComb (C : Set E) (nonempty : C.Nonempty)
    (coneComb_mem : ∀ x ∈ C, ∀ y ∈ C, ∀ a : R, 0 ≤ a → ∀ b : R, 0 ≤ b → a • x + b • y ∈ C) :
     PointedCone R E where
```